### PR TITLE
[PLAYER-2911]

### PR DESCRIPTION
**Pull request by cacel**
Ooyala Heartbeat for SSAI

### CODE REVIEW

File: js/ssai_heartbeat.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    function startHeartBeat() {
+      stopHeartBeat();
+
+      heartbeatTimer = setInterval(reportHeartBeat, config.Interval);
+    }
+
+    function reportHeartBeat() {
+        if(reportingPaused){
```
Comment (dallen42) [inline js/ssai_heartbeat.js]: Indentation seems is a bit too deep here, should be only two spaces instead of four

File: js/ssai_heartbeat.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      log('initialization completed', config);
+    }
+
+    function stopHeartBeat() {
+      clearInterval(heartbeatTimer);
+    }
+
+    function startHeartBeat() {
+      stopHeartBeat();
+
```
Comment (dallen42) [inline js/ssai_heartbeat.js]: @cacel remove extra line here

File: js/ssai_heartbeat.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      //events subscribe
+      mb.subscribe(OO.EVENTS.VC_WILL_PLAY, 'heartbeat', _onVcWillPlay);
+      mb.subscribe(OO.EVENTS.EMBED_CODE_CHANGED, 'heartbeat', _onEmbedCodeChanged);
+      mb.subscribe(OO.EVENTS.PLAYHEAD_TIME_CHANGED, 'heartbeat', _onPlayheadTimeChange);
+      mb.subscribe(OO.EVENTS.PAUSE, 'heartbeat', _onPause);
+      mb.subscribe(OO.EVENTS.PLAY, 'heartbeat', _onPlay);
+      mb.subscribe(OO.EVENTS.VC_PLAYED, 'hearbeat', _onPlayed);
+      mb.subscribe(OO.EVENTS.DESTROY, 'heartbeat', destroy)
+
```
Comment (dallen42) [inline js/ssai_heartbeat.js]: @cacel could remove these extra lines as well

File: js/ssai_heartbeat.js (hunk 1 above)
Comment (cacel) [reply js/ssai_heartbeat.js]: done

File: js/ssai_heartbeat.js (hunk 2 above)
Comment (cacel) [reply js/ssai_heartbeat.js]: done

File: js/ssai_heartbeat.js (hunk 3 above)
Comment (cacel) [reply js/ssai_heartbeat.js]: done

File: js/ssai_heartbeat.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    function reportHeartBeat() {
+        if(reportingPaused){
+          return;
+        }
+
+        var reportingURL = config.ReportingPathPattern.replace(/<hostname>/g, hostname).replace(/<embed_code>/g, embed_code).replace(/<ssai_guid>/g, ssai_guid);
+        var data = {
+          playheadpos: parseInt(playheadPosition),
+          pingfrequency: parseInt(config.Interval / 1000)
```
[the remark this answers is not in the document]
Comment (cacel) [reply js/ssai_heartbeat.js]: According to the API (http://adassets.ooyala.com/api/index.html#/vod/setPbPosVOD), the expected data is playheadpos and pingfrequency.

File: js/ssai_heartbeat.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+      if ( result.length && result[1] ) {
+        return result[1];
+      }
+      return '';
+    }
+
+    function _onVcWillPlay(event, videoId, url) {
+      streamURL = url || '';
+      ssai_guid = parseGUID(streamURL);
```
[the remark this answers is not in the document]
Comment (cacel) [reply js/ssai_heartbeat.js]: done

File: js/ssai_heartbeat.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+      var result = reg.exec(url);
+
+      if ( result.length && result[1] ) {
+        return result[1];
+      }
+      return '';
+    }
+
+    function _onVcWillPlay(event, videoId, url) {
+      streamURL = url || '';
```
[the remark this answers is not in the document]
Comment (cacel) [reply js/ssai_heartbeat.js]: done

File: js/ssai_heartbeat.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      }
+    }
+
+    function _onPlayheadTimeChange(event, currentTime, duration) {
+      movieDuration = duration || 0;
+      playheadPosition = currentTime || 0;
+    }
+
+    function _onEmbedCodeChanged(event, embedCode){
+      embed_code = embedCode || '';
```
[the remark this answers is not in the document]
Comment (cacel) [reply js/ssai_heartbeat.js]: done

File: js/ssai_heartbeat.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      stopHeartBeat();
+      heartbeatTimer = setInterval(reportHeartBeat, config.Interval);
+    }
+
+    function reportHeartBeat() {
+      if(reportingPaused){
+        return;
+      }
+
+      var reportingURL = config.ReportingPathPattern.replace(/<hostname>/g, hostname).replace(/<embed_code>/g, embedCode).replace(/<ssai_guid>/g, ssaiGuid);
```
Comment (pilievOoyala) [inline js/ssai_heartbeat.js]: should be reportingUrl

File: js/ssai_heartbeat.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        contentType: 'text/plain',
+        dataType: 'text',
+        success:function(){
+          log('Heartbeat was sent successfully');
+        }
+      });
+
+    }
+
+    function parseGUID(url) {
```
Comment (pilievOoyala) [inline js/ssai_heartbeat.js]: should be parseGuid

**Comment by pilievOoyala** [comment]
Two more minor style changes otherwise LGTM. You can merge after that

**Pull request merged**